doc/event-alarm-framework/events-producer.md

```diff
@@ -0,0 +1,212 @@
+</br>
+
+## Design
+
+### YANG schema
+1. Schema defines the description of the event, its unique tag and all the possible parameters.
+2. Schema can be maintained in multiple files, preferably one per process/continer/host.
+3. All the schema files are copied into one single location (e.g. /usr/shared/sonic/events).
+4. The schema for processes running in host are copied into this location at image creation/install time.
+5. The schema for processes running inside the containers are held inside the containers and copied into the shared location on the first run. This allows for independent container upgrade scenarios.
+6. NB clients could use the schema to understand/analyze the events
+
+#### A sample Defintion
+A sample:
+```
+module events-bgp {
+    . . .
+    container bgp_status {
+        list event_list {
+            key "type ip"
+            
+            leaf type {
+                enum "admin_up";
+                enum "admin_down";
+                enum "idle";
+                enum "active";
+                enum "open";
+                enum "established";
+            }
+            
+            leaf ip {
+                type inet:ip-address;
+                description "IP of neighbor";
+            }
+            
+            leaf timestamp {
+                type inet::date-and-time;
+                description "time of the event"
+            }
+        }
+    }
+``` 
+
+### Event detection
+The event detection could happen in many ways
+- Update the code to directly invoke Event reporter, which will stream it out.
+- Run a periodic check (via monit/cron/daemon/...) to identify an event and raise it via event reporter
+- Watch syslog messages for specific pattern; Upon match parse the required data out and call event reporter.
+- Any other means of getting alerted on an event
+
+#### Log message based detection
+At high level:
+1. This is a two step process.
+2. The process raising the event sends a sylog message out.
+3. A watcher scans all the syslog messages emitted and parse/check for events of interest.
+4. When matching message arrives, raise the event
+
+Here you have code that sends the log and a watcher who has the regex pattern for that log message to match. Anytime the log messsage is changed the pattern has to be updated for the event to fire consistently across releases.
+
+Though this sounds like a redundant/roundabout way, this helps as below.
+- For III party code, not owned by SONiC, this is an acceptable solution
+- For code that SONiC own, a code update would take in the order of months to reach thousands of switches, but this approach can be applied instantly to every switch
+- A new event addition could be as simple as copying couple of small files to a switch and a rsyslog/container restart.
+
+##### Design at high level
+- Configure a rsyslog plugin as per process with rsyslog.d.
```

Comment:
>per process [](http://example.com/codeflow?start=32&length=11)

How about one plugin per container? If 'per process', the syslog will be duplicated for multiple plugins.

Reply:
No. The rsyslog.d  will spawn one plugin per process and pass on only messages from that process to that plugin.

Reply:
Can you clarify more on how this works? What changes would we have to make to create these plugins?